accounts-db/src/hardened_unpack.rs

```diff
@@ -112,27 +112,28 @@ where
         // first by ourselves when there are odd paths like including `..` or /
         // for our clearer pattern matching reasoning:
         //   https://docs.rs/tar/0.4.26/src/tar/entry.rs.html#371
-        let parts = path.components().map(|p| match p {
-            CurDir => Some("."),
-            Normal(c) => c.to_str(),
-            _ => None, // Prefix (for Windows) and RootDir are forbidden
-        });
+        let parts = path
+            .components()
+            .map(|p| match p {
+                CurDir => Ok("."),
+                Normal(c) => c.to_str().ok_or(()),
+                _ => Err(()), // Prefix (for Windows) and RootDir are forbidden
+            })
+            .collect::<std::result::Result<Vec<_>, _>>();
 
         // Reject old-style BSD directory entries that aren't explicitly tagged as directories
         let legacy_dir_entry =
             entry.header().as_ustar().is_none() && entry.path_bytes().ends_with(b"/");
         let kind = entry.header().entry_type();
         let reject_legacy_dir_entry = legacy_dir_entry && (kind != Directory);
 
-        if parts.clone().any(|p| p.is_none()) || reject_legacy_dir_entry {
+        if parts.is_err() || reject_legacy_dir_entry {
             return Err(UnpackError::Archive(format!(
                 "invalid path found: {path_str:?}"
             )));
         }
+        let parts = parts.unwrap();
```

Comment:
I think we can replace manual variant check + unwrap with a match, something like:
```rust
        let parts = match (parts, reject_legacy_dir_entry) {
            (Ok(parts), false) => parts,
            _ => {
                return Err(UnpackError::Archive(format!(
                    "invalid path found: {path_str:?}"
                )));
            }
        };
```